### PR TITLE
Properties: specific properties

### PR DESCRIPTION
**Pull request by pascal-brand-st-dev**

### CODE REVIEW

File: lib/libutee/tee_api_property.c [hunk 1]
```diff
@@ -219,8 +219,17 @@ TEE_Result TEE_GetPropertyAsString(TEE_PropSetHandle propsetOrEnumerator,
...
 	if (res != TEE_SUCCESS) {
-		if (res == TEE_ERROR_SHORT_BUFFER)
+		if (res == TEE_ERROR_SHORT_BUFFER) {
+			if (type == USER_TA_PROP_TYPE_BINARY_BLOCK) {
+				/*
+				 * in this case, we must enlarge the buffer
+				 * with the size of the of the base64 encoded
+				 * see base64_enc() function
+				 */
+				tmp_len = 4 * ((tmp_len + 2) / 3) + 1;
```
Comment (jenswi-linaro) [inline lib/libutee/tee_api_property.c]: This calculation would be nice keep in a macro in `base64.h` or so.

File: core/tee/tee_svc.c [hunk 2]
```diff
@@ -437,23 +424,40 @@ const struct tee_props tee_propset_tee[] = {
 	 */
 };
 
+__weak const struct tee_props *vendor_propset_client = NULL;
+__weak const unsigned long vendor_propset_client_len = 0;
+__weak const struct tee_props *vendor_propset_ta = NULL;
+__weak const unsigned long vendor_propset_ta_len = 0;
+__weak const struct tee_props *vendor_propset_tee = NULL;
```
Comment (jenswi-linaro) [inline core/tee/tee_svc.c]: I would prefer a struct containing both the pointer and the length like:
```
struct tee_vendor_props {
    const struct tee_props *props;
    size_t len;
};
```
To guarantee that both pointer and length is overridden at the same time.
By the way, there's no need to assign `0` and `NULL` that's done automatically.

File: core/arch/arm/plat-vexpress/main.c [hunk 3]
```diff
@@ -138,3 +138,71 @@ void console_flush(void)
...
+	size_t *bin_len = (size_t *)bin;
+	uint8_t *bin_val = (uint8_t *)(&bin[1]);
+
+	if (*blen < sizeof(bin)) {
+		*blen = sizeof(bin);
+		return TEE_ERROR_SHORT_BUFFER;
+	}
+	*blen = sizeof(bin);
+
+	memcpy(&data[0], &sess->ctx->uuid, sizeof(TEE_UUID));
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-vexpress/main.c]: just `data` should be enough.

File: core/arch/arm/plat-vexpress/main.c [hunk 4]
```diff
@@ -138,3 +138,71 @@ void console_flush(void)
 {
 	pl011_flush(CONSOLE_UART_BASE);
 }
+
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-vexpress/main.c]: This stuff should go in a separate file.

File: core/arch/arm/plat-vexpress/main.c [hunk 5]
```diff
@@ -138,3 +138,71 @@ void console_flush(void)
...
+	if (tee_otp_get_die_id(&data[sizeof(TEE_UUID)],
+			       ta_endorsement_seed_size))
+		return TEE_ERROR_BAD_STATE;
+
+	res = tee_hash_createdigest(TEE_ALG_SHA256, data, sizeof(data),
+				    bin_val, ta_endorsement_seed_size);
+	if (res != TEE_SUCCESS)
+		return TEE_ERROR_BAD_STATE;
+
+	*bin_len = ta_endorsement_seed_size;
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-vexpress/main.c]: Someone begging for 64-bit trouble?

File: core/arch/arm/plat-vexpress/main.c [hunk 6]
```diff
@@ -138,3 +138,71 @@ void console_flush(void)
...
+ *
+ * The output buffer is the "binary" struct defined in
+ * the "prop_value" union and therefore comprises:
+ * -  4 bytes: the size of the binary value data (32)
+ * - 32 bytes: the binary value data (endorsement seed)
+ *
+ * Note that this code assumes an endorsement seed
+ * size == device ID size for convenience.
+ */
+static const size_t ta_endorsement_seed_size = 32;
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-vexpress/main.c]: This doesn't have to be a global variable, a define or a local variable would be preferable.

File: core/arch/arm/plat-vexpress/main.c [hunk 7]
```diff
@@ -138,3 +138,71 @@ void console_flush(void)
...
+				    bin_val, ta_endorsement_seed_size);
+	if (res != TEE_SUCCESS)
+		return TEE_ERROR_BAD_STATE;
+
+	*bin_len = ta_endorsement_seed_size;
+
+	return tee_svc_copy_to_user(sess, (void *)buf, bin, sizeof(bin));
+}
+
+const struct tee_props vendor_propset_array_tee[] = {
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-vexpress/main.c]: Can be static

File: core/tee/tee_svc.c [hunk 8]
```diff
@@ -437,23 +424,40 @@ const struct tee_props tee_propset_tee[] = {
 	 */
 };
 
+__weak const struct tee_props *vendor_propset_client = NULL;
+__weak const unsigned long vendor_propset_client_len = 0;
```
Comment (jenswi-linaro) [inline core/tee/tee_svc.c]: We usually use `size_t` instead of `unsigned long` for these things.

**Comment by pascal-brand-st-dev** [comment]
Updated

**Comment by pascal-brand-st-dev** [comment]
@paswan-ms here is an implementation of your original submission https://github.com/OP-TEE/optee_os/pull/460. Would that be ok?

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/tee/tee_svc.c [hunk 9]
```diff
@@ -601,7 +610,14 @@ TEE_Result syscall_get_property_name_to_index(unsigned long prop_set,
 		if (!strcmp(kname, props[i].name)) {
 			res = tee_svc_copy_to_user(sess, index, &i,
 						   sizeof(*index));
-			break;
+			goto out;
+		}
+	}
+	for (i = size; i < size + vendor_size; i++) {
```
Comment (paswan-ms) [inline core/tee/tee_svc.c]: Why not for (I = 0 ; I < size ; I++) and then vendor_props[I] ?

**Comment by paswan-ms** [comment]
Looks good, thanks!

File: core/tee/tee_svc.c (hunk 9 above)
Comment (pascal-brand-st-dev) [inline core/tee/tee_svc.c]: if so, we have to modify `tee_svc_copy_to_user(... &i ...)` below

Comment (paswan-ms) [inline core/tee/tee_svc.c]: OK, got it.

**Pull request merged**